ext/strscan/strscan.c

```diff
@@ -1266,6 +1268,55 @@ strscan_peep(VALUE self, VALUE vlen)
     return strscan_peek(self, vlen);
 }
 
+/*
+ * call-seq:
+ *   scan_integer
+ *
+ * Equivalent to #scan with a [+-]?\d+ pattern, and returns an Integer or nil.
+ *
+ * The scanned string must be encoded with an ASCII compatible encoding, otherwise
+ * Encoding::CompatibilityError will be raised.
+ */
+static VALUE
+strscan_scan_integer(VALUE self)
+{
+    char *ptr, *buffer;
+    long len = 0;
+    VALUE buffer_v, integer;
+    struct strscanner *p;
+
+    GET_SCANNER(self, p);
+    CLEAR_MATCH_STATUS(p);
+
+    rb_must_asciicompat(p->str);
+
+    ptr = CURPTR(p);
+
+    if (ptr[len] == '-' || ptr[len] == '+') {
+        len++;
+    }
+
+    if (!isdigit(ptr[len])) {
+        return Qnil;
+    }
+
+    MATCHED(p);
+    p->prev = p->curr;
+
+    while(isdigit(ptr[len])) {
+        len++;
+    }
+
+    buffer = ALLOCV_N(char, buffer_v, len + 1);
+
+    MEMCPY(buffer, CURPTR(p), char, len);
+    buffer[len] = '\0';
+    integer = rb_cstr2inum(buffer, 10);
```

Comment:
Ah, we want to avoid this copy...

Reply:
OK. We can use `rb_cstr2inum()` for now. Could you propose the fast path to Ruby itself? It may be useful for extensions.